### PR TITLE
New syntax for simple and shared pointers + Templated Functions

### LINKED ISSUE #14: Template support for free functions in the wrapper
## Feature

Template support for free functions in wrapper

## Motivation

Functions like triangulatePoint3, which can take different calibrations as arguments, currently require multiple copy-paste operations to add wrapper support for different calibrations.  

## Pitch

Adding template support for free functions allows gtsam to better be DRY.

### PR DESCRIPTION
**Pull request by varunagrawal**
This PR updates the syntax for wrapping simple and shared pointers.

E.g. class:

```cpp
class Test {
 public:
  void takePointer(double* d) const {}
  void shareTheLove(boost::shared_ptr<gtsam::Pose3> pose) const {}
};
```

To wrap the above class, we simply do:

```cpp
class Test {
  void takePointer(double@ d) const;
  void shareTheLove(gtsam::Pose3* pose) const;
};
```

and we are done!!

This PR also adds support for templated functions and fixes #14.
We can now do the below in the wrapper:

```cpp
template<T={double}>
void InsertFunStuff(gtsam::Values@ values, int key, const T& fun_stuff);
```

### CODE REVIEW

**Comment by ProfFan** [comment]
Unit test is not updated

**Comment by varunagrawal** [comment]
Woke up and realized I was missing the unit tests.

**Comment by varunagrawal** [comment]
@ProfFan I added/updated the unit tests.

**Comment by varunagrawal** [comment]
@ProfFan @dellaert added the unit tests.

File: gtwrap/interface_parser.py [hunk 1]
```diff
@@ -730,24 +775,26 @@ class GlobalFunction:
...
                  name: str,
                  return_type: ReturnType,
                  args_list: ArgumentList,
+                 template: Template,
                  parent: str = ''):
         self.name = name
         self.return_type = return_type
         self.args = args_list
-        self.is_const = None
+        self.template = template
```
Comment (dellaert) [inline gtwrap/interface_parser.py:797]: What happened to const? Deciding to no longer keep track of const is a different PR/decision (with pros and cons)

**Review by dellaert — COMMENTED** [review]
Wow, that seemed like a lot of work! 
A merge needs to be coordinated with an almost simultaneous GTSAM and GTDynamics PR, right?

File: gtwrap/interface_parser.py (hunk 1 above)
Comment (varunagrawal) [reply gtwrap/interface_parser.py:797]: It's never used in this class, isn't initialized in the constructor and a global function with a const at the end aka `void GlobalFunc() const;` is something I've never seen.

**Comment by varunagrawal** [comment]
Not really. If we merge this and pull the latest version, the wrapped code will need to updated for GTDynamics which should be easy-peasy.

GTSAM requires us to update the git subtree since we include `wrap` within the repo. That can be a whole PR in itself and won't affect GTSAM's current working.

**Comment by varunagrawal** [comment]
It wasn't a lot of manual work as it was mental work of figuring out all the dependencies, adding comments, and making sure the parsing rules were correct (`pyparsing`'s docs could use some minor improvements).

**Comment by varunagrawal** [comment]
@dellaert done

File: tests/geometry.h [hunk 2]
```diff
@@ -104,7 +104,6 @@ class Test {
 };
 
 pair<gtsam::NonlinearFactorGraph**, gtsam::Values**> load2D(string filename, Test** model, int maxID, bool addNoise, bool smart);
-pair<gtsam::NonlinearFactorGraph**, gtsam::Values**> load2D(string filename, gtsam::noiseModel::Diagonal** model, int maxID, bool addNoise, bool smart);
```
Comment (ProfFan) [inline tests/geometry.h]: why is this deleted

**Review by ProfFan — APPROVED** [review]

**Review by dellaert — APPROVED** [review]
Please check GTSAM wrapper does not break.

File: tests/geometry.h (hunk 2 above)
Comment (varunagrawal) [reply tests/geometry.h]: It was causing all the indices in the matlab wrapper to be incremented, which was a mess. :confused:

**Comment by varunagrawal** [comment]
Wrapper works for both GTSAM and GTDynamics. Just needed to update all the `double&` to `double` since I fixed the parsing rule for that (this was primarily an issue in `gtsam_unstable.i`.
